src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2222,12 +2222,44 @@ public override BoundNode VisitWhileStatement(BoundWhileStatement node)
             return base.VisitWhileStatement(node);
         }
 
-        public override BoundNode VisitWithExpression(BoundWithExpression expr)
+        public override BoundNode VisitWithExpression(BoundWithExpression withExpr)
         {
-            // PROTOTYPE: This is wrong
-            SetResultType(expr,
-                expr.CloneMethod?.ReturnTypeWithAnnotations.ToTypeWithState()
-                    ?? TypeWithState.Create(expr.Type, NullableFlowState.NotNull));
+            Debug.Assert(!IsConditionalState);
+
+            var receiver = withExpr.Receiver;
+            VisitRvalue(receiver);
+            _ = CheckPossibleNullReceiver(receiver, checkNullableValueType: false);
+
+            var resultSlot = GetOrCreatePlaceholderSlot(withExpr);
+            // carry over the null state of members of 'receiver' to the result of the with-expression.
+            TrackNullableStateForAssignment(receiver, ResultType.ToTypeWithAnnotations(), resultSlot, ResultType, MakeSlot(receiver));
+
+            foreach (var (leftSymbol, right) in withExpr.Arguments)
+            {
+                if (leftSymbol is PropertySymbol prop)
+                {
+                    // PROTOTYPE: attributes on positional record parameters do not propagate
+                    // to the synthesized properties. Perhaps this is by design?
+                    // If we want them to propagate then perhaps we should move
+                    // members such as SourcePropertySymbol.HasAllowNull, etc. to SourceOrRecordPropertySymbol.
+                    var leftAnnotations = GetPropertySetterAnnotations(prop);
+                    var leftLValueType = ApplyLValueAnnotations(prop.TypeWithAnnotations, leftAnnotations);
+
+                    var rightState = VisitOptionalImplicitConversion(right, targetTypeOpt: leftLValueType, useLegacyWarnings: false, trackMembers: true, AssignmentKind.Assignment);
+                    CheckDisallowedNullAssignment(rightState, leftAnnotations, right.Syntax.Location);
+                    var propSlot = GetOrCreateSlot(prop, resultSlot);
+                    TrackNullableStateForAssignment(right, leftLValueType, propSlot, rightState, MakeSlot(right));
+
+                    // PROTOTYPE: Should we AdjustSetValue for properties?
```

Comment:
It appears that init-only properties can be manually implemented, so yes, we should AdjustSetValue.

https://github.com/dotnet/roslyn/blob/891fd421c4488d861c4959bbc91245eb4b118a06/src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs#L646